### PR TITLE
JDK 21 support, major determinism improvements, lots of upgrades

### PR DESCRIPTION
**Pull request by jjudd**
This should get us
* Builds working on JDK 21
* Deterministic builds pretty much everywhere. I haven't found any other bugs related to the rules, but I wouldn't be surprised if another bug or two is lurking.
* Upgrades on a lot of our third party dependencies.

### CODE REVIEW

File: src/main/scala/higherkindness/rules_scala/workers/zinc/compile/ZincRunner.scala [hunk 1]
```diff
@@ -271,17 +278,42 @@ object ZincRunner extends WorkerMain[Namespace] {
...
-    analysisStore.set(AnalysisContents.create(compileResult.analysis, compileResult.setup))
+    val path = analysisStoreFile.getCanonicalPath()
+    val analysisStoreText = AnalysisUtil.getAnalysisStore(
+      new File(path.substring(0, path.length() - 3) + ".text.gz"),
+      true,
+      usePersistence,
+    )
+    // Filter out libraryClassNames from the analysis because it is non-deterministic.
+    // Can stop doing this once the bug in Zinc is fixed. Check the comment on FilteredRelations
+    // for more info.
```
Comment (gregghz) [inline src/main/scala/higherkindness/rules_scala/workers/zinc/compile/ZincRunner.scala:289]: is there a bug filed against zinc for this?

File: src/main/scala/higherkindness/rules_scala/workers/zinc/compile/FilteredInfos.scala [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import xsbti.{Problem, VirtualFileRef}
+
+/**
+ * The consistent analysis format in Zinc does not send the Position in the Problems in the SourceInfos through the read
+ * or write mapper, so the file paths are absolute rather than relativized, which means they are not machine independent
+ * and reproducible.
+ *
+ * This filters out the paths from those Problems/Positions, so we can get a deterministic output.
+ *
+ * TODO: fix this problem in Zinc.
```
Comment (gregghz) [inline src/main/scala/higherkindness/rules_scala/workers/zinc/compile/FilteredInfos.scala:14]: Is there a bug to reference?

File: src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala [hunk 3]
```diff
@@ -89,7 +89,7 @@ trait WorkerMain[S] {
                 System.err.println(s"Uncaught exception in Future while proccessing WorkRequest $requestId:")
                 e.printStackTrace(System.err)
               }
-            }
+            }(scala.concurrent.ExecutionContext.global)
```
Comment (gregghz) [inline src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala:92]: do you have any info about the differences between these two execution contexts? You mention "IO tasks" being handled by this default context. Would it make sense to use a cached/unbounded context? Or would that cause memory issues by the amount of threads we could end up with?

File: tests/format/scalafmt/test [hunk 4]
```diff
@@ -23,36 +23,36 @@ bazel run :unformatted-binary-copy.format
...
+# # non-default output format test: library
+# ## check
+# bazel run :formatted-library.format-test
+# ! bazel run :unformatted-library.format-test
+# ## copy
+# cp unformatted.template.scala unformatted.tmp.scala
+# ! bazel run :unformatted-library-copy.format-test
+# bazel run :unformatted-library-copy.format
+# bazel run :unformatted-library-copy.format-test
+# rm unformatted.tmp.scala
```
Comment (gregghz) [inline tests/format/scalafmt/test]: are all these things commented out on purpose?

**Review by gregghz — APPROVED** [review]
I've looked over everything and it looks good to me. I'll approve but with the caveat that this is really the first time I've looked closely at rules_scala -- we probably should get a second opinion.

Looks great though, thanks for all this work.

File: rules/private/phases/phase_zinc_compile.bzl [hunk 5]
```diff
@@ -72,7 +73,7 @@ def phase_zinc_compile(ctx, g):
         ] + [zinc.deps_files for zinc in zincs],
     )
 
-    outputs = [g.classpaths.jar, mains_file, analysis_store, used, tmp]
+    outputs = [g.classpaths.jar, mains_file, analysis_store, analysis_store_text, used, tmp]
```
Comment (BrianXDong) [inline rules/private/phases/phase_zinc_compile.bzl:76]: Do we know how expensive this is to do?
Additionally, do we have a visible place to put a TODO to remove this logging once we've figured out the reproducibility problems?

File: src/main/scala/higherkindness/rules_scala/workers/zinc/compile/ZincRunner.scala (hunk 1 above)
Comment (jjudd) [reply src/main/scala/higherkindness/rules_scala/workers/zinc/compile/ZincRunner.scala:289]: Not yet, but I intend to fix it when I have a moment. Hopefully sometime in the next couple months.

File: src/main/scala/higherkindness/rules_scala/workers/zinc/compile/FilteredInfos.scala (hunk 2 above)
Comment (jjudd) [reply src/main/scala/higherkindness/rules_scala/workers/zinc/compile/FilteredInfos.scala:14]: Not yet. I also intend to fix this one in the next couple months.

File: src/main/scala/higherkindness/rules_scala/workers/zinc/compile/ZincRunner.scala [hunk 6]
```diff
@@ -79,7 +79,7 @@ object ZincRunner extends WorkerMain[Namespace] {
   protected[this] def work(worker: Namespace, args: Array[String], out: PrintStream) = {
     val usePersistence: Boolean = worker.getBoolean("use_persistence") match {
       case p: java.lang.Boolean => p
-      case null                 => true
+      case null                 => false
```
Comment (BrianXDong) [inline src/main/scala/higherkindness/rules_scala/workers/zinc/compile/ZincRunner.scala:82]: > At some point we should tie this option into the ZincConfiguration
parameters in Bazel land.

Anywhere where we can file this? 

Additionally, iirc we're comfortable with disabling this due to how RBE handles actions, where in our current state persistence isn't useful, but after persistent workers we may want persistence, right?
If so, we should also make sure that that is documented somewhere to do.

File: tests/format/scalafmt/test (hunk 4 above)
Comment (jjudd) [reply tests/format/scalafmt/test]: Nope. Pretty sure that's my bad. I'll go fix this.

File: rules/private/phases/phase_zinc_compile.bzl (hunk 5 above)
Comment (jjudd) [reply rules/private/phases/phase_zinc_compile.bzl:76]: I don't think it's too expensive to do, but don't have numbers to back that up. I didn't notice any difference with it on or off, but I also wasn't taking measurements.

I've got a reminder set for first week of September to come disable this if we no longer need it.

File: tests/format/scalafmt/test (hunk 4 above)
Comment (jjudd) [reply tests/format/scalafmt/test]: I've uncommented this locally and the test is passing. I'll push it up with the other requested changes.

**Review by BrianXDong — APPROVED** [review]
Similar boat to @gregghz where this is the first time I've deeply looked at rules_scala, but if you're confident with your testing, then I'm comfortable as well.

I think a conversation on whether our RBE cluster is ready to handle the additional load before enabling scala in RBE would be prudent.

Thanks for doing all of this work!

File: src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala (hunk 3 above)
Comment (jjudd) [reply src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala:92]: I think that this is a pretty lightweight task. It's just writing to stdout so the worker can communicate with the bazel server, but it's doing so synchronized. I figure we might as well free up the part of the worker that's actually doing work to do that work and let the synchronized IO part happen in another thread pool.

File: src/main/scala/higherkindness/rules_scala/workers/zinc/compile/ZincRunner.scala (hunk 6 above)
Comment (jjudd) [reply src/main/scala/higherkindness/rules_scala/workers/zinc/compile/ZincRunner.scala:82]: Currently this is just filed in my brain. I started work on this in this PR with
```
if zinc_configuration.incremental:
        execution_requirements_tags["no-sandbox"] = "1"
```

In `phase_zinc_compile.bzl`. I just need to extend it to do more than that.

I plan to keep making improvements, but the PR was already so large I just wanted to get a PR up before making any more changes.

**Pull request merged**